File: vol2/vol2-java-examples/src/main/java/com/heatonresearch/aifh/evolutionary/opp/CompoundOperator.java
/*
 * Artificial Intelligence for Humans
 * Volume 2: Nature Inspired Algorithms
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh
 *
 * Copyright 2014 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */
package com.heatonresearch.aifh.evolutionary.opp;

import com.heatonresearch.aifh.evolutionary.genome.Genome;
import com.heatonresearch.aifh.evolutionary.train.EvolutionaryAlgorithm;
import com.heatonresearch.aifh.general.collections.ObjectHolder;
import com.heatonresearch.aifh.randomize.GenerateRandom;

/**
 * A compound operator randomly chooses sub-operators to perform the actual
 * operation. Each of the sub-operators can be provided with a weighting.
 */
public class CompoundOperator implements EvolutionaryOperator {

    /**
     * The owner of this operator.
     */
    private EvolutionaryAlgorithm owner;

    /**
     * The sub-operators that make up this compound operator.
     */
    private final OperationList components = new OperationList();

    /**
     * @return the components
     */
    public OperationList getComponents() {
        return this.components;
    }

    /**
     * @return the owner
     */
    public EvolutionaryAlgorithm getOwner() {
        return this.owner;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void init(final EvolutionaryAlgorithm theOwner) {
        this.owner = theOwner;
        for (final ObjectHolder<EvolutionaryOperator> obj : this.components
                .getList()) {
            obj.getObj().init(theOwner);
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int offspringProduced() {
        return this.components.maxOffspring();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int parentsNeeded() {
        return this.components.maxOffspring();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void performOperation(final GenerateRandom rnd, final Genome[] parents,
                                 final int parentIndex, final Genome[] offspring,
                                 final int offspringIndex) {
        final EvolutionaryOperator opp = this.components.pick(rnd);
        opp.performOperation(rnd, parents, parentIndex, offspring,
                offspringIndex);
    }
}

File: vol2/vol2-csharp-examples/AIFH-Vol2-Core/Core/Evolutionary/Opp/CompoundOperator.cs
// Artificial Intelligence for Humans
// Volume 2: Nature-Inspired Algorithms
// C# Version
// http://www.aifh.org
// http://www.jeffheaton.com
//
// Code repository:
// https://github.com/jeffheaton/aifh
//
// Copyright 2014 by Jeff Heaton
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// For more information on Heaton Research copyrights, licenses
// and trademarks visit:
// http://www.heatonresearch.com/copyright
//
using AIFH_Vol2.Core.Evolutionary.Genome;
using AIFH_Vol2.Core.Evolutionary.Train;
using AIFH_Vol2.Core.General.Collections;
using AIFH_Vol2.Core.Randomize;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AIFH_Vol2.Core.Evolutionary.Opp
{
    /// <summary>
    /// A compound operator randomly chooses sub-operators to perform the actual
    /// operation. Each of the sub-operators can be provided with a weighting.
    /// </summary>
    public class CompoundOperator : IEvolutionaryOperator
    {
        /// <summary>
        /// The owner of this operator.
        /// </summary>
        private IEvolutionaryAlgorithm _owner;

        /// <summary>
        /// The sub-operators that make up this compound operator.
        /// </summary>
        private OperationList _components = new OperationList();

        /// <summary>
        /// The components.
        /// </summary>
        public OperationList Components
        {
            get
            {
                return _components;
            }
        }

        /// <summary>
        /// The owner.
        /// </summary>
        public IEvolutionaryAlgorithm Owner
        {
            get
            {
                return _owner;
            }
        }

        /// <inheritdoc/>
        public void Init(IEvolutionaryAlgorithm theOwner)
        {
            _owner = theOwner;
            foreach (ObjectHolder<IEvolutionaryOperator> obj in _components
                    .ObjList)
            {
                obj.Obj.Init(theOwner);
            }
        }

        /// <inheritdoc/>
        public int OffspringProduced
        {
            get
            {
                return _components.MaxOffspring;
            }
        }

        /// <inheritdoc/>
        public int ParentsNeeded
        {
            get
            {
                return _components.MaxOffspring;
            }
        }

        /// <inheritdoc/>
        public void PerformOperation(IGenerateRandom rnd, IGenome[] parents,
                                     int parentIndex, IGenome[] offspring,
                                     int offspringIndex)
        {
            IEvolutionaryOperator opp = _components.Pick(rnd);
            opp.PerformOperation(rnd, parents, parentIndex, offspring,
                    offspringIndex);
        }
    }
}
